src/percy-agent-client/dom.ts

```diff
@@ -0,0 +1,213 @@
+
+    // Sometimes you'll want to transform the DOM provided into one ready for snapshotting
+    // For example, if your test suite runs tests in an element inside a page that
+    // lists all yours tests. You'll want to "hoist" the contents of the testing container to be
+    // the full page. Using a dom transformation is how you'd achieve that.
+    if (this.options.domTransformation) {
+      try {
+        dom = this.options.domTransformation(dom)
+      } catch (error) {
+        console.error('Could not transform the dom: ', error)
+      }
+    }
+
+    return doctype + dom.outerHTML
+  }
+
+  private getDoctype(): string {
+    return this.clonedDOM.doctype
+      ? this.doctypeToString(this.clonedDOM.doctype)
+      : this.defaultDoctype
+  }
+
+  private doctypeToString(doctype: DocumentType): string {
+    const publicDeclaration = doctype.publicId
+      ? ` PUBLIC "${doctype.publicId}" `
+      : ''
+    const systemDeclaration = doctype.systemId
+      ? ` SYSTEM "${doctype.systemId}" `
+      : ''
+
+    return (
+      `<!DOCTYPE ${doctype.name}` + publicDeclaration + systemDeclaration + '>'
+    )
+  }
+
+  /**
+   * Takes the raw DOM from the snapshot, performs necessary mutations,
+   * and finally returns a deep clone of the original DOM.
+   *
+   */
+  private cloneDOM(): HTMLDocument {
+    this.mutateOriginalDOM()
+    // Any because its type changes when given to `stabilizeDOM`
+    let clone = this.originalDOM.cloneNode(true) as any
+    clone = this.stabilizeDOM(clone)
+
+    return clone
+  }
+
+  /**
+   * Serialize parts of the DOM that aren't preserved in a deep clone. Anything
+   * that is not encoded in the DOM tree needs to be accounted for here (CSSOM,
+   * input values, canvas, etc).
+   *
+   * This method should always capture these values from the original DOM and
+   * serialize into the cloned DOM. Never mutate the original DOM.
+   *
+   */
+  private stabilizeDOM(clonedDOM: HTMLDocument): HTMLElement {
+    let stabilizedDOMClone = this.serializeInputElements(clonedDOM)
+
+    // We only want to serialize the CSSOM if JS isn't enabled.
+    if (!this.options.enableJavaScript) {
+      stabilizedDOMClone = this.serializeCSSOM(stabilizedDOMClone)
+    }
+
+    return stabilizedDOMClone.documentElement
+  }
+
+  /**
+   * Capture in-memory form element values and serialize those values to their
+   * respective elements in the cloned DOM.
+   *
+   * Form element values for non-controlled elements won't be captured with a
+   * domClone. To fix this, we explicitly set the `value` attribute on those
+   * form controls to make sure they persist in snapshots.
+   *
+   */
+  private serializeInputElements(clonedDOM: HTMLDocument): HTMLDocument {
+    const formNodes = this.originalDOM.querySelectorAll('input, textarea')
+    const formElements = Array.from(formNodes)
+
+    formElements.forEach((elem: any) => {
+      const inputId = elem.attributes['data-percy-element-id'].value
```

Comment:
```suggestion
      const inputId = elem.getAttribute('data-percy-element-id')
```

I think this is comparable?

Reply:
Yeah that's a real good one 👍